PcapDotNet/src/PcapDotNet.Packets/Dhcp/DhcpDatagram.cs

```diff
@@ -0,0 +1,368 @@
+        {
+            get
+            {
+                ParseOptions();
+                return _options;
+            }
+        }
+
+        internal DhcpDatagram(byte[] buffer, int offset, int length) : base(buffer, offset, length)
+        {
+        }
+
+        private void ParseServerHostName()
+        {
+            if (_serverHostName == null)
+            {
+                //at the moment we only interpret server host name as sname and ignore options
+                byte[] byteServerHostName = ReadBytes(Offset.Sname, 64);
+                _serverHostName = Encoding.ASCII.GetString(byteServerHostName).TrimEnd('\0');
+            }
+        }
+
+        private void ParseBootFileName()
+        {
+            if (_bootFileName == null)
+            {
+                //at the moment we only interpret server host name as sname and ignore options
+                byte[] byteBootFileName = ReadBytes(Offset.File, 128);
+                _bootFileName = Encoding.ASCII.GetString(byteBootFileName).TrimEnd('\0');
+            }
+        }
+
+        private void ParseOptions()
+        {
+            if (_options == null)
+            {
+                List<DhcpOption> options = new List<DhcpOption>();
+                int pos = IsDhcp ? Offset.OptionsWithMagicCookie : Offset.Options;
+                while (pos < Length)
+                {
+                    options.Add(DhcpOption.CreateInstance(this, ref pos));
+                }
+                this._options = new ReadOnlyCollection<DhcpOption>(options);
+            }
+        }
+
+        /// <summary>
+        /// Creates a Layer that represents the datagram to be used with PacketBuilder.
+        /// </summary>
+        public override ILayer ExtractLayer()
+        {
+            return new DhcpLayer
+            {
+                MessageType = MessageType,
+                HardwareType = HardwareType,
+                HardwareAddressLength = HardwareAddressLength,
+                Hops = Hops,
+                TransactionId = TransactionId,
+                SecondsElapsed = SecondsElapsed,
+                DhcpFlags = Flags,
+                ClientIpAddress = ClientIpAddress,
+                YourClientIpAddress = YourClientIpAddress,
+                NextServerIpAddress = NextServerIpAddress,
+                RelayAgentIpAddress = RelayAgentIpAddress,
+                ClientHardwareAddress = ClientHardwareAddress,
+                ServerHostName = ServerHostName,
+                BootFileName = BootFileName,
+                IsDhcp = IsDhcp,
+                Options = Options.ToList()
+            };
+        }
+
+        internal static int GetLength(bool isDhcp, IList<DhcpOption> options)
+        {
+            int length = isDhcp ? Offset.OptionsWithMagicCookie : Offset.Options;
+
+            if (options != null)
+            {
+                length += options.Sum(p => 1 + (!(p is DhcpPadOption || p is DhcpEndOption) ? 1 : 0) + p.Length); //Type + Len? + Option
```

Comment:
Replace "p" with "option".
Replace "1" with "sizeof(byte)".
Can you rely on p.MessageType instead of its class type?